src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -175,11 +185,44 @@ public Task StopAsync()
             catch (Exception ex)
             {
                 Log.UnexpectedExceptionInStop(Logger, ConnectionId, ex);
-            }
-            
+            }
+
             return Task.CompletedTask;
         }
 
+        public async Task CloseAsync()
+        {
+            if (Interlocked.CompareExchange(ref _closeStage, ClosingSendingFin1, ClosingDefault) != ClosingDefault)
+            {
+                // CloseAsync is in progress or done.
+                return;
+            }
+
+            while (Interlocked.CompareExchange(ref _closeStage, ClosingWaitingApplicationTasks, ClosingReceivedFinAck1) != ClosingReceivedFinAck1)
```

Comment:
Why are we doing this instead of waiting on the task?

Reply:
Because we won't stop `incomingTask` in the newest proposal (we let our user doing this by themselves, or close connections on the `ASRS` side after connection migration feature has been implemented)

Reply:
This code seems unnecessarily complex. What's the new flow?